Precise framing

elalish opened the pull request:
Fixes #731 
Fixes #735 
Fixes #524 

This updates our framing to fit our models precisely within a window, meaning for all camera orbit values and all window aspect ratios the model does not clip, but it is also tight, so the model will touch the boundary of the window at least twice. This also makes the camera distance (and therefore the perspective) fixed with respect to window aspect ratio, which means it is now possible for us to calculate seamless poster transitions: #599 

It does take a few extra milliseconds, but it only roughly doubles the time we were already spending on calculating the bounding box. I've also made it easy for us to add some extra margin to this framing if that's desired; I'd like feedback on that. It might even make sense as an attribute...

elalish commented on index.html:
```diff
@@ -159,17 +159,18 @@ <h3 class="grouping-title">Attributes</h3>
+              $distance", like "10deg 75deg 1.5m". Also supports units in
+              radians ("rad") for angles and centimeters ("cm") or millimeters
+              ("mm") for camera distance. Camera distance can also be set as a
+              percentage ('%'), where 100% gives the model tight framing within
```
It's not relative to 'auto'; 'auto' = 105%, as stated below.

elalish commented on src/features/controls.ts:
```diff
@@ -324,38 +324,40 @@ export const ControlsMixin = <T extends Constructor<ModelViewerElementBase>>(
-      // and will only end up showing the skysphere if zoomed out enough
-      const minimumRadius = near + framedHeight / 2.0;
-      const maximumRadius = this[$idealCameraDistance]!;
+      if (this.fieldOfView === DEFAULT_FIELD_OF_VIEW) {
```
This is the auto case, in which case we change the fov based on aspect ratio and such. If you set fov, then this fanciness doesn't make sense.

cdata replied:
Oh, I missed that `DEFAULT_FIELD_OF_VIEW` is `auto` and not a fixed value. SG.

elalish commented on src/test/fidelity/config-reader.ts:
```diff
@@ -24,7 +31,15 @@ export class ConfigReader {
-        return scenario;
+        const output = Object.assign({}, defaultScenario, scenario);
+
+        // This is necessary because Object.assign is not deep
```
Yeah, I was less than thrilled when I discovered what Object.assign actually does.

elalish commented on src/three-components/Model.ts:
```diff
@@ -29,16 +30,19 @@ const $currentScene = Symbol('currentScene');
   public boundingBox = new Box3();
   public size = new Vector3();
+  public idealCameraDistance = 0;
+  public fovAspect = 0;
```
I added more docs than were in this file before! What's the rule here? Or do you just mean we should be adding to docs to all our files that don't currently include them?

elalish commented on src/three-components/Model.ts:
```diff
@@ -262,10 +273,43 @@ export default class Model extends Object3D {
+          value, radiusXZ / (this.idealCameraDistance - Math.abs(vertex.y)));
+    };
+    this.fovAspect =
+        reduceVertices(this.modelContainer, horizontalFov) / verticalFov;
```
Correct, which is why I moved it from ModelScene to Model. Now that this function does not depend on canvas aspect ratio, it is only calculated when the model changes, not when resize occurs.

elalish commented on src/three-components/ModelUtils.ts:
```diff
@@ -155,4 +156,51 @@ export const moveChildren = (from: Object3D, to: Object3D) => {
+            let geometry = object.geometry;
+
+            if (geometry !== undefined) {
+              if (geometry.isGeometry) {
```
No idea; just copying three for robustness. Seems like even if it doesn't now, it could in the future.

cdata replied:
Arguably the other geometries are only there for legacy purposes (they are rather expensive to use), but it's fine if this is a straight-up copy of the Three.js traversal algorithm.

elalish commented on src/three-components/SmoothControls.ts:
```diff
@@ -47,8 +47,8 @@ export interface SmoothControlsOptions {
-  minimumPolarAngle: Math.PI / 8,
-  maximumPolarAngle: Math.PI - Math.PI / 8,
+  minimumPolarAngle: 0,        // Math.PI / 8,
+  maximumPolarAngle: Math.PI,  // - Math.PI / 8,
```
Good question; I don't have a strong opinion on this one, but I found it really annoying that I couldn't look down on a model. I can put it back if you want, but was there a good reason for these defaults?

cdata replied:
I think the rational at the time was two fold: 1) when scrolling, it was annoying / slightly ugly how easy it was to accidentally orbit to the top-down view unintentionally 2) elastic extents would look better if the defaults weren't looking directly up / down on the model.

Personally I don't feel that strongly about the user-facing defaults as long as the extents are configurable in public API... but they aren't yet 🤷‍♂

elalish commented on test/fidelity/renderers/model-viewer/index.html:
```diff
@@ -69,6 +67,10 @@
+      const {theta, phi, radius} = scenario.orbit;
+      modelViewer.cameraOrbit = `${theta}rad ${phi}deg ${radius}m`;
+      const {x, y, z} = scenario.target;
+      modelViewer.cameraTarget = `${x}m ${y}m ${z}m`;
```
Yes, I'm trying too now.

elalish commented on src/conversions.ts:
```diff
@@ -25,30 +25,36 @@ import {parseValues, ValueNode} from './parsers.js';
+        case 'mm':
+          absolute = value / 1000;
+          break;
+        case '%':
```
Good call, will fix.

elalish commented on src/features/controls.ts:
```diff
@@ -243,41 +246,38 @@ export const ControlsMixin = <T extends Constructor<ModelViewerElementBase>>(
     }
 
     [$updateCameraOrbit]() {
       let sphericalValues = deserializeSpherical(this.cameraOrbit);
```
SG

cdata commented on src/conversions.ts:
```diff
@@ -86,26 +86,29 @@ const convertAngleValueNode =
-    };
+    (sphericalString: string,
+     defaultValues: [number, number, number|null, number|null]):
+        [number, number, number|null, number|null] => {
```
Is the returned factor ever considered outside of this helper, aside from immediately interpreting the result? Perhaps we can pre-apply the factor to the returned radius? That way we can stick to returning three not-null numbers always.

Also, let's document the fourth default in the comment above.

elalish replied:
Unfortunately I couldn't find a way to do that. The two numbers need to be used in different ways (only one of them is ever valid). I toyed with returning something that said how to use that number instead. I could use the sign as a flag, but figured that's pretty opaque. Do you have a better way?

elalish commented:
Oddly, some of the filament renders line up exactly with the model-viewer ones, and others don't. I'm wondering if some transform is or isn't applied in Filament's world space. Still, I don't think it's worth blocking us on, as we had some trouble with this before anyway.

cdata commented:
I have added that as of this PR, substantially all of the issues described in #524 appear to be fixed. The misreported `modelIsVisible` value can no longer be reproduced, and the bug reported by @pushmatrix in that thread also can no longer be reproduced. I believe the resolution came from correcting an improperly computed aspect ratio.

`modelIsVisible` remains an incongruous and misleading property, and I have filed https://github.com/GoogleWebComponents/model-viewer/issues/773 to cover its redesign.

cdata commented:
Also a note to add that I still have additional work to verify that #735 is fixed as of this change.

cdata commented:
I confirmed that #735 is also fixed by this change.

elalish commented:
Okay, I've got all the fidelity tests lining up properly now, as well as fixing issues around null cameraOrbits and such. I also updated SmoothControls so that it now works even when radius is set to zero. I also changed the default minRadius to zero, since I think that is more logical (allows easy camera placement if you're not really orbiting) and it was part of fixing the fidelity test alignment.

pushmatrix commented:
@cdata Hmm, I still seem to be getting the issue from #524.

I pulled this branch and built the js.

I have it up and running here: https://congruous-freesia.glitch.me/

cdata commented:
@pushmatrix can you share more details for repro? I tried out your test case in Chrome / OSX and it appears to cycle between two rendered models as I push the button. Although one weird thing is that the first model does not appear to be correctly framed... Is that a detail of the model, or an unexpected bug?

![framing-fix](https://user-images.githubusercontent.com/240083/65888265-16282b00-e354-11e9-9aff-1dba2a8ea85f.gif)

pushmatrix commented:
@cdata What version of Chrome are you on? I'm on 76.0.3809.132.  My repro steps are just pressing "Show viewer 2", and the model doesn't load.

However I do notice that it wants me to update, so maybe that will fix it. But then I'd be confused why a slightly outdated Chrome causes this.

I tested in Safari and it works.

pushmatrix commented:
RE: Model being off center, that is a bug. The model is centered at the base of the stroller.

cdata commented:
@pushmatrix given that description, I might guess that something broken is sticky in your cache.

@elalish looking at the framing issue with the first model in https://congruous-freesia.glitch.me/, it seems framed correctly until it starts to autorotate. This bug appears to affect `master` but the latest release (`v0.6.2`) appears to be unaffected. I'm going to file a separate issue for this.

cdata commented:
(I was already on Chrome 77. I will see if I can find an easy way to try out 76, but we are pretty certain that the bug was just bad JavaScript math on our part so hopefully browser version isn't a factor 🙏 )

cdata commented:
Filed https://github.com/GoogleWebComponents/model-viewer/issues/784 to track the autorotate / framing issue.

pushmatrix commented:
@cdata I just used Browserstack to test out 76 and the bug is still there. In 77 it disappears. Hmmmmmmmm

cdata commented:
@pushmatrix okay thanks, I'll look into it

cdata commented:
@pushmatrix confirmed repro in 76 :(

cdata commented:
Sigh. The problem in 76 goes away if I force the canvas onto its own render layer. In other words, this is why we can't have nice things.

cdata commented:
@pushmatrix @elalish please see https://github.com/GoogleWebComponents/model-viewer/pull/785 for a proposed remedy to the Chrome 76 issue!

pushmatrix commented:
Big props @elalish !
